delphi/translators/for2py/genModFileLog.py

```diff
@@ -152,14 +159,211 @@ def populate_mappers(file_path, file_to_mod_mapper, mod_to_file_mapper,
         file_to_mod_mapper[file_path] = module_names_lowered.copy()
         file_to_mod_mapper[file_path].append(get_file_last_modified_time(
             file_path))
-
+        if (
+            ("end subroutine" in file_content.lower()
+                or "endsubroutine" in file_content.lower())
+            or ("end function" in file_content.lower()
+                    or "endfunction" in file_content.lower())
+        ):
+            # Bring the pointer back to the first character position of a file
+            f.seek(f_pos)
+            # regex for module, subroutine, and function
+            # Because Fortran allows two different syntax for end (i.e. end module and
+            # endmodule), we need to check both cases.
+            modu_regex = re.compile('\s*module (?P<module>(?P<name>(\w*)\n))')
+            end_modu_regex = re.compile('\s*end module (?P<module>(?P<name>(\w*)\n))')
+            endmodu_regex = re.compile('\s*endmodule (?P<module>(?P<name>(\w*)\n))')
+            
+            interface_regex  = re.compile('\s*interface (?P<interface>(?P<name>(\w*)\n))')
+            end_intr_regex = re.compile('\s*end interface\n')
+            endintr_regex = re.compile('\s*endinterface\n')
+
+            subr_regex = re.compile('\s*subroutine (?P<subroutine>(?P<name>.*?)\((?P<args>.*)\))')
+            end_subr_regex = re.compile('\s*end subroutine (?P<subroutine>(?P<name>(\w*)\n))')
+            endsubr_regex = re.compile('\s*endsubroutine (?P<subroutine>(?P<name>(\w*)\n))')
+            
+            func_regex = re.compile('\s*(?P<type>(double precision|double|float|int|integer|logical|real|str|string))\sfunction (?P<function>(?P<name>.*?)\((?P<args>.*)\))')
+            end_func_regex = re.compile('\s*end function (?P<function>(?P<name>(\w*)\n))')
+            endfunc_regex = re.compile('\s*endfunction (?P<function>(?P<name>(\w*)\n))')
+
+            var_dec_regex = re.compile('\s*(?P<type>(double|float|int|integer|logical|real|str|string)),?\s(::)*(?P<variables>(.*))\n')
+            dev_type_regex = re.compile('\s*type\s*\((?P<type>.*)\)\s(?P<variables>(.*))')
+            char_type_regex = re.compile('\s*(character\*(\(\*\)|\d*))\s*(?P<variables>.*)')
+
+            dtype_regex = re.compile('\s*type (?P<type>(?P<name>(\w*)\n))')
+            end_dtype_regex = re.compile('\s*end type (?P<type>(?P<name>(\w*)\n))')
+            enddtype_regex = re.compile('\s*endtype (?P<type>(?P<name>(\w*)\n))')
+
+            current_modu = None
+            current_subr = None
+            current_intr = None
+            current_func = None
+
+            isProcedure =False
+
+            line = f.readline().lower()
+            while (line):
+                #  Removing any inline comments
```

Comment:
@skdebray Unfortunately, `genModLogFile.py` runs before any other program and runs only once to scan all raw Fortran source code files before preprocessed, so I had. add this line of code to manually remove inline comments.

Reply:
@hlim1 Thanks for the clarification.  In addition to removing comments, the preprocessor also merges continuation lines.  Are continuation lines an issue for you at all?

In general, I worry about replicating the same functionality in multiple places in the code -- this is usually not a good idea for a variety of reasons.  So, unless there are reasons to avoid it, my preference here would be to invoke the preprocessor and then work with the preprocessed code.

Reply:
@skdebray I've modified the code to preprocess the file before begin to analyze and collect. module information, so any inline comments and continued lines are handled.
One thing that I would like to mention about the change in `preprocessor.py` is that I added a case to skip processing include files.
```
209     if not forModLogGen:
210         lines = process_includes(lines, infile)
```
This is because I got an error at line 223 `with open(infile, mode="r", encoding="latin-1") as f:` with an error message saying
that the program is trying to open a file that does not exist, such as `delphi/tests/data/program_analysis/COMSWI.blk`.
Before adding this condition to skip, I added a file existence checking condition `os.path.isfile(infile)` at line 169 to only process if file exist else return an empty string.
However, this caused an infinite loop in the preprocessor.py under process_include function, I guess. 
```
149 def process_includes(lines, infile):
150     """ process_includes() processes INCLUDE statements, which behave like
151         the #include preprocessor directive in C.
152     """
153     chg = True
154     while chg:
155         chg = False
156         include_idxs = [i for i in range(len(lines))
157                           if line_is_include(lines[i]) is not None]
158
159         # include_idxs is a list of the index positions of INCLUDE statements.
160         # Each such statement is processed by replacing it with the contents
161         # of the file it mentions.  We process include_idxs in reverse so that
162         # processing an INCLUDE statement does not change the index position of
163         # any remaining INCLUDE statements.
164         for idx in reversed(include_idxs):
165             chg = True
166             include_f = line_is_include(lines[idx])
167             assert include_f is not None
168             include_path = path_to_target(infile, include_f)
169             incl_lines = preprocess_file(include_path)
170             lines = lines[:idx] + incl_lines + lines[idx+1:]
171
172     return lines
```
Therefore, I instead decided to skip processing the included files only when it's for module summary generation as it is not important.

Reply:
@hlim1 That's an interesting bug in the preprocessor that I had not expected.  Thanks for catching and diagnosing it.  I'm surprised the solution you tried (preprocess the included file only if it exists) didn't work -- that seems like a very natural and very reasonable solution.  I think that, ultimately, the right solution would be for me to fix the preprocessor to deal with this situation correctly.  But we have a lot of things to do, so let's move forward with your implementation decision.

Let me know when you've finished updating the code, at that point I'll go ahead and merge.